File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.py
import sys

from collections import deque

# 말 움직임 8가지, 말은 장애물 뛰어넘을 수 있음
# 원숭이는 상하좌우만 가능, 또한 말처럼 총 k번만 이동 가능
# -> 상태가 있음 -> 상태가 있다는 것은 3차원..

k = int(input())
m, n = map(int, input().split())  # 가로, 세로
g = [list(map(int, input().split())) for _ in range(n)]

# 0인 빈칸, 1은 벽, 벽이 있는 곳으로 이동 불가
blank = 0
wall = 1

# (0,0) -> (n-1,m-1) 로 이동해야함 -> 최단거리

startX, startY = 0, 0
endX, endY = n - 1, m - 1

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]
hx = [-2, -2, -1, 1, 2, 2, 1, -1]
hy = [-1, 1, 2, 2, 1, -1, -2, -2]


def isInner(x, y):
    if 0 <= x < n and 0 <= y < m:
        return True
    return False


def BFS(startX, startY):
    global res
    dq = deque()
    dq.append((startX, startY, 0, 0))  # 시작점, 이동거리, 말로 움직인 횟수
    INF = int(1e9)
    # [n][m][k]
    dis = [[[INF] * (k+1) for _ in range(m)] for _ in range(n)]  # [n][m][k] 만듬
    dis[startX][startY][0] = 0  # 시작점 초기화

    while dq:
        x, y, d, cnt = dq.popleft()
        if (x,y) == (endX,endY):
            res = min(res,d)
            break
        # 먼저 원숭이 움직임
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if not isInner(nx, ny): continue
            if g[nx][ny] == blank:  # 이동 가능
                if d + 1 < dis[nx][ny][cnt]:  # 갱신 가능
                    dis[nx][ny][cnt] = d + 1
                    dq.append((nx, ny, d + 1, cnt))
        # 말 움직임
        if cnt < k:  # 아직 말로 이동 가능
            for i in range(8):
                nx = x + hx[i]
                ny = y + hy[i]
                if not isInner(nx, ny): continue
                if g[nx][ny] == blank:
                    if d + 1 < dis[nx][ny][cnt + 1]:
                        dis[nx][ny][cnt + 1] = d + 1
                        dq.append((nx,ny,d+1,cnt+1))


res = int(1e9)
BFS(startX,startY)
if res == int(1e9):
    print(-1)
else:
    print(res)

File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;
import java.util.StringTokenizer;
import java.util.function.IntFunction;

public class Main {

    // ! 말이 되고픈 원숭이
    // ! 원숭이 : 상하좌우 움직임
    // ! 말 : 나이트 움직임
    // ! 총 k번만 이동 가능. 그 외에는 인접한 칸으로만 움직임 가능
    // ! 격자판의 맨 왼쪽 위에서 맨 오른쪽 아래까지 이동. - > 최단거리
    // ! 0은 이동 가능, 1은 이동 불가.
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    static int n, m, k; // ! k는 말처럼 이동 가능 횟수
    static int[][] g;
    static int[][][] dis; // ! 최단 거리 기록용 배열
    static int endX,endY,INF = 10000000;
    static int MIN_DIS;

    static int[] dx = {-1, 0, 1, 0};
    static int[] dy = {0, 1, 0, -1};
    static int[] hx = {-2, -1, 1, 2, 2, 1, -1, -2};
    static int[] hy = {1, 2, 2, 1, -1, -2, -2, -1};

    private static class Data {

        int x, y, dis, cnt;

        Data(int x, int y, int dis, int cnt) {
            this.x = x;
            this.y = y;
            this.dis = dis;
            this.cnt = cnt;
        }
    }

    public static boolean isInner(int x, int y){
        return x >= 0 && x < n && y >= 0 && y < m;
    }

    public static void main(String[] args) throws Exception {
        st = new StringTokenizer(br.readLine());
        k = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        m = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());

        g = new int[n][m];
        MIN_DIS = INF;
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                g[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        dis = new int[n][m][k+1]; // ! k번까지 가능
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                Arrays.fill(dis[i][j], INF);
            }
        }

        endX = n-1;
        endY = m-1;
        BFS(0,0);
        if (MIN_DIS == INF) {
            System.out.println(-1);
        } else {
            System.out.println(MIN_DIS);
        }
    }

    public static void BFS(int startX,int startY) {
        Deque<Data> dq = new ArrayDeque<>();
        dq.offer(new Data(startX, startY, 0, 0));
        dis[startX][startY][0] = 0; // ! 최단거리 기록하고 시작

        while (!dq.isEmpty()) {
            Data now = dq.poll();
            if (now.x == endX && now.y == endY) {
                MIN_DIS = Math.min(MIN_DIS, now.dis);
                continue;
            }

            // ! 원숭이 움직임 먼저
            for (int i = 0; i < 4; i++) {
                int nx = now.x + dx[i];
                int ny = now.y + dy[i];
                if(!isInner(nx,ny)) // ! 외부 좌표라면 패스
                    continue;

                if (g[nx][ny] == 0) {
                    if (now.dis + 1 < dis[nx][ny][now.cnt]) {
                        dis[nx][ny][now.cnt] = now.dis + 1;
                        dq.offer(new Data(nx, ny, now.dis + 1, now.cnt));
                    }
                }
            }

            // ! 말 움직임 먼저
            for (int i = 0; i < 8; i++) {
                int nx = now.x + hx[i];
                int ny = now.y + hy[i];
                if(!isInner(nx,ny))
                    continue;

                if (g[nx][ny] == 0 && now.cnt < k) {
                    if(now.dis + 1 < dis[nx][ny][now.cnt + 1]){
                        dis[nx][ny][now.cnt + 1] = now.dis + 1;
                        dq.offer(new Data(nx, ny, now.dis + 1, now.cnt + 1));
                    }
                }
            }
        }
    }
}